Support dependencies from registries for artifact dependencies, take 2

fee1-dead opened the pull request:
<!-- homu-ignore:start -->
<!--
NOTICE: Due to limited review capacity, the Cargo team is not accepting new
features or major changes at this time. Please consult with the team before
opening a new PR. Only issues that have been explicitly marked as accepted
will be reviewed.

Thanks for submitting a pull request 🎉! Here are some tips for you:

* If this is your first contribution, read "Cargo Contribution Guide":
  https://doc.crates.io/contrib/
* Run `cargo fmt --all` to format your code changes.
* Small commits and pull requests are always preferable and easy to review.
* If your idea is large and needs feedback from the community, read how:
  https://doc.crates.io/contrib/process/#working-on-large-features
* Cargo takes care of compatibility. Read our design principles:
  https://doc.crates.io/contrib/design.html
* When changing help text of cargo commands, follow the steps to generate docs:
  https://github.com/rust-lang/cargo/tree/master/src/doc#building-the-man-pages
* If your PR is not finished, set it as "draft" PR or add "WIP" in its title.
* It's ok to use the CI resources to test your PR, but please don't abuse them.

### What does this PR try to resolve?

Explain the motivation behind this change.
A clear overview along with an in-depth explanation are helpful.

You can use `Fixes #<issue number>` to associate this PR to an existing issue.

### How should we test and review this PR?

Demonstrate how you test this change and guide reviewers through your PR.
With a smooth review process, a pull request usually gets reviewed quicker.

If you don't know how to write and run your tests, please read the guide:
https://doc.crates.io/contrib/tests

### Additional information

Other information you want to mention in this PR, such as prior arts,
future extensions, an unresolved problem, or a TODO list.
-->
<!-- homu-ignore:end -->

This is a continuation of #12062, and closes #10405.

r? @ehuss

Here are things this PR changes:

1. Add information about artifact dependencies to the index. This bumps index_v to 3 for people using bindeps.
2. Make `RustcTargetData` mutable for the feature resolver. This is so that we can query rustc for target info as late as resolving features, as that is when we really know if a package is really going to be used.

bors commented:
:umbrella: The latest upstream changes (presumably #12426) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

ehuss commented:
@fee1-dead Just wanted to check in and acknowledge that I have seen this PR, but I haven't had the time to look at it. I will try to get to it soon.

ehuss commented on src/cargo/sources/registry/index.rs:
```diff
@@ -807,7 +815,7 @@ impl<'a> SummariesCache<'a> {
             .ok_or_else(|| anyhow::anyhow!("cache expected 4 bytes for index schema version"))?;
         let index_v = u32::from_le_bytes(index_v_bytes.try_into().unwrap());
-        if index_v != INDEX_V_MAX {
+        if index_v != INDEX_V_MAX && index_v != 3 {
```
Can you say more about why this check was added? It doesn't look like v3 is being written in `SummariesCache::serialize` anyway, so it is always 2.

I'm thinking this isn't necessary (yet), because older cargos don't fail to parse the new entries. The key part of this index check is `IndexSummary::parse`. If we make a change that causes that to fail, then older cargos will ignore those entries, and not place them in the cache (it assumes they are from a future format it does not understand). However, since bindeps is just adding new fields to the JSON, that doesn't cause a problem (new fields are ignored). 

When we stabilize, I think we'll need to set INDEX_V_MAX to 3, but I'm not sure it is necessary to make any changes at this time.

fee1-dead replied:
I think this was simply carried over from the last PR. I don't remember my reasoning for this and I have removed this condition.

ehuss commented:
Thanks, this looks great! 

Can you add a test to make sure the `v` index version filtering works? I think that would look like something like this:

```rust
Package::new("bar", "1.0.0").publish();
Package::new("bar", "1.0.1")
    .schema_version(3)
    .add_dep(dep.artifact("bin", Some(target.to_string())))
    .publish();

// Verify that without `-Zbindeps` that it does not use 1.0.1.
let p = project()
    .file("Cargo.toml",
        r#"
            [package]
            name = "foo"
            version = "0.1.0"

            [dependency]
            bar = "1.0"
        "#)
    .file("src/lib.rs", "")
    .build();

p.cargo("tree").with_stdout("...").run();

// And with -Zbindeps it can use 1.0.1.
p.cargo("update -Zbindeps").with_stderr("...").run();

// And without -Zbindeps, now that 1.0.1 is in Cargo.lock, it should fail.
p.cargo("check").with_status(101).with_stderr("...").run();
```

I think that should roughly cover it, but if you have other ideas on how to try to break it, I think it would be good to make sure that has plenty of coverage. 

----

I noticed that this doesn't update `download_accessible`. Were there problems with that, or is that something that we can iterate on later? (Not a blocker for this PR.)

----

Related to `download_accessible`, I think `FeatureResolver::is_proc_macro` will panic if it needs to process a proc-macro loaded from an artifact dependency with a target, and there is some problem downloading (like a network error). That is, something like the following will panic:

```rust
#[cargo_test]
fn proc_macro_in_artifact_dep() {
    // Forcing FeatureResolver to check a proc-macro for a dependency behind a
    // target dependency.
    if cross_compile::disabled() {
        return;
    }
    Package::new("pm", "1.0.0")
        .file("src/lib.rs", "")
        .file(
            "Cargo.toml",
            r#"
                [package]
                name = "pm"
                version = "1.0.0"

                [lib]
                proc-macro = true

            "#,
        )
        .publish();
    let alternate = cross_compile::alternate();
    Package::new("bin-uses-pm", "1.0.0")
        .target_dep("pm", "1.0", alternate)
        .file("src/main.rs", "fn main() {}")
        .publish();
    // Simulate a network error downloading the proc-macro.
    std::fs::remove_file(cargo_test_support::paths::root().join("dl/pm/1.0.0/download")).unwrap();
    let p = project()
        .file(
            "Cargo.toml",
            &format!(
                r#"
                [package]
                name = "foo"
                version = "0.1.0"
                edition = "2021"

                [dependencies]
                bin-uses-pm = {{ version = "1.0", artifact = "bin", target = "{alternate}"}}
            "#
            ),
        )
        .file("src/lib.rs", "")
        .build();

    p.cargo("check -Z bindeps")
        .masquerade_as_nightly_cargo(&["bindeps"])
        .with_stderr("")
        .run();
}
```

----

Can you make sure to document `IndexPackage::v` with the new version 3 changes?

fee1-dead commented on src/cargo/core/resolver/features.rs:
```diff
@@ -852,28 +864,49 @@ impl<'a, 'cfg> FeatureResolver<'a, 'cfg> {
+                                // and per-pkg-targets are not immediately known.
+                                let mut activate_target = |target| {
+                                    self.target_data
+                                        .merge_compile_kind(CompileKind::Target(target))
```
Done.

fee1-dead commented on tests/testsuite/artifact_dep.rs:
```diff
@@ -1445,13 +1445,7 @@ foo v0.0.0 ([CWD])
-//       For reference, see comments by ehuss https://github.com/rust-lang/cargo/pull/9992#discussion_r801086315 and
-//       joshtriplett https://github.com/rust-lang/cargo/pull/9992#issuecomment-1033394197 .
 #[cargo_test]
-#[ignore = "broken, need artifact info in index"]
```
Added

fee1-dead commented on tests/testsuite/artifact_dep.rs:
```diff
@@ -2894,8 +2896,7 @@ fn check_transitive_artifact_dependency_with_different_target() {
         .with_stderr_contains(
-            "error: could not find specification for target `custom-target`.\n  \
-            Dependency `baz v0.0.0 [..]` requires to build for target `custom-target`.",
+            "error: failed to run `rustc` to learn about target-specific information",
```
Fixed

fee1-dead commented on src/cargo/sources/registry/index.rs:
```diff
@@ -178,7 +178,7 @@ struct Summaries {
 enum MaybeIndexSummary {
     /// A summary which has not been parsed, The `start` and `end` are pointers
-    /// into [`Summaries::raw_data`] which this is an entry of.
+    /// into [`Summaries::raw_data`] which this isRegistryDependency an entry of.
```
fixed

fee1-dead commented:
> Can you add a test to make sure the `v` index version filtering works?

added

> I noticed that this doesn't update `download_accessible`. Were there problems with that, or is that something that we can iterate on later? (Not a blocker for this PR.)

Yeah, let's iterate on this later, I specifically did not look into updating `download_accessible` because it turned out to be a bit complicated for the last PR. I've added your test case as an ignored test.

> Can you make sure to document `IndexPackage::v` with the new version 3 changes?

Yep, done.

fee1-dead commented:
not sure why check-version-bump is failing for `cargo-credential`. This PR doesn't modify that crate

bors commented:
:umbrella: The latest upstream changes (presumably #12381) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

ehuss commented:
> not sure why check-version-bump is failing for cargo-credential. This PR doesn't modify that crate

I think there was an issue that was since fixed (#12508).

ehuss commented:
Thanks, looks good!  I appreciate you working on this.

I filed #12554 for the download_accessible issue, and #12555 to track updating crates.io (though I'm not sure when we should actually do that).

Can you resolve the merge conflict? Unfortunately it looks like I don't have permission to push to your branch.

ehuss commented:
Thanks!

@bors r+

bors commented:
:pushpin: Commit 43dccc7535f62c84ea2dc32a250608ff9bc4fa0b has been approved by `ehuss`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/cargo) for this repository.

<!-- @bors r=ehuss 43dccc7535f62c84ea2dc32a250608ff9bc4fa0b -->
<!-- homu: {"type":"Approved","sha":"43dccc7535f62c84ea2dc32a250608ff9bc4fa0b","approver":"ehuss","queue":"https://bors.rust-lang.org/homu/queue/cargo"} -->

bors commented:
:hourglass: Testing commit 43dccc7535f62c84ea2dc32a250608ff9bc4fa0b with merge e2c6a3e0e83ce6075d1fdf491102c5b8dc121fb3...
<!-- homu: {"type":"BuildStarted","head_sha":"43dccc7535f62c84ea2dc32a250608ff9bc4fa0b","merge_sha":"e2c6a3e0e83ce6075d1fdf491102c5b8dc121fb3"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/cargo/actions/runs/5971757211/job/16201912999)
Approved by: ehuss
Pushing e2c6a3e0e83ce6075d1fdf491102c5b8dc121fb3 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"ehuss","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/cargo/actions/runs/5971757211/job/16201912999"},"merge_sha":"e2c6a3e0e83ce6075d1fdf491102c5b8dc121fb3"} -->

elchukc commented:
> I think `FeatureResolver::is_proc_macro` will panic if it needs to process a proc-macro loaded from an artifact dependency with a target, and there is some problem downloading (like a network error).

Working on a fix for this test right now. Is this an example of a spurious network error, for which the desired behaviour that it retries?

elchukc commented:
> I think `FeatureResolver::is_proc_macro` will panic if it needs to process a proc-macro loaded from an artifact dependency with a target, and there is some problem downloading (like a network error).

@ehuss  Working on a fix for this test right now. Is this an example of a spurious network error, for which the desired behaviour that it retries?

ehuss commented:
@elchukc Sorry, I'm not quite understanding your question. There is a test (already checked in) called `proc_macro_in_artifact_dep`. It demonstrates what happens if there is any kind of network error when downloading a proc-macro (it currently panics). The fix should make it so that doesn't panic, but instead provide a reasonable error message.
